### PR TITLE
[DeepSeek V3.2] Enable trtllm NSA with bf16 kvcache

### PR DESCRIPTION
**Pull request by akhilg-nv**
## Motivation

Enables NSA backend with trtllm kernels for sparse attention. This can be more efficient than FlashMLA when the head size isn't a multiple of 64 and hence requires padding. This PR enables with BF16 KVCache, FP8 will follow in another PR.

## Modifications

This change interfaces with the new kernel added in flashinfer to use trtllm kernel for decode in NSA backend. There are also modifications made to `server_args.py` to enable the new option.

## Accuracy Tests

```
python -m sglang.launch_server --model-path deepseek-ai/DeepSeek-V3.2 --tp 8 --dp 8 --enable-dp-attention --nsa-decode-backend trtllm --kv-cache-dtype bfloat16 --reasoning-parser deepseek-v3 --trust-remote-code

python3 -m sglang.test.run_eval --port 30000 --eval-name gpqa --num-examples 198 --max-tokens 120000 --repeat 8 --thinking-mode deepseek-v3 |& tee logs/trtllm_mla_sparse_gpqa.log

Repeat: 8, mean: 0.853
Scores: ['0.854', '0.854', '0.859', '0.879', '0.848', '0.838', '0.859', '0.833']
```

Testing MTP `--speculative-algorithm EAGLE --speculative-num-steps 2 --speculative-eagle-topk 1 --speculative-num-draft-tokens 3`

GPQA:
```
Repeat: 8, mean: 0.84898 [1:19:07<04:19, 23.58s/it]
Scores: ['0.838', '0.854', '0.869', '0.818', '0.848', '0.843', '0.874', '0.838']

Avg accept len: 2.3740
Min accept len: 1.9800
Max accept len: 2.8500

Avg accept rate: 0.7913
Min accept rate: 0.6600
Max accept rate: 0.9500
```

LongBenchV2:
`python -m sglang.test.run_eval \
    --eval-name longbench_v2 \
    --host 127.0.0.1 \
    --port 30000 \
    --model deepseek-ai/DeepSeek-V3.2 \
    --max-context-length 128000 \
    --max-tokens 16384 \
    --num-threads 16
`

```
Total latency: 1313.224 s
Score: 0.550

Avg accept len: 2.5248
Min accept len: 2.0000
Max accept len: 2.9800

Avg accept rate: 0.8419
Min accept rate: 0.6700
Max accept rate: 0.9900
```

## Benchmarking and Profiling

Trace shows the new kernel call 
(about 14 microseconds):

image

image

Trace with the FlashMLA path with bf16 KVCache shows the FlashMLA sparse kernel is slower (about 26.5 microseconds):

image

image

Comparing FlashMLA sparse and TRTLLM (both with bf16 kvcache) shows similar average results for decode.

`python3 -m sglang.bench_serving --model deepseek-ai/DeepSeek-V3.2 --warmup-requests 5 --max-concurrency 16 --num-prompts 80 --random-range-ratio 0.8 --random-input-len 4096 --random-output-len 4096 --profile --profile-by-stage` 

FlashMLA Sparse (BF16 KVCache)
```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 16        
Successful requests:                     80        
Benchmark duration (s):                  57.26     
Total input tokens:                      29016     
Total input text tokens:                 29016     
Total generated tokens:                  17012     
Total generated tokens (retokenized):    16921     
Request throughput (req/s):              1.40      
Input token throughput (tok/s):          506.74    
Output token throughput (tok/s):         297.10    
Peak output token throughput (tok/s):    576.00    
Peak concurrent requests:                21        
Total token throughput (tok/s):          803.83    
Concurrency:                             10.87     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   7780.15   
Median E2E Latency (ms):                 5522.24   
P90 E2E Latency (ms):                    17340.74  
P99 E2E Latency (ms):                    33557.02  
---------------Time to First Token----------------
Mean TTFT (ms):                          348.70    
Median TTFT (ms):                        276.23    
P99 TTFT (ms):                           683.36    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          34.78     
Median TPOT (ms):                        36.56     
P99 TPOT (ms):                           57.35     
---------------Inter-Token Latency----------------
Mean ITL (ms):                           35.21     
Median ITL (ms):                         28.07     
P95 ITL (ms):                            151.14    
P99 ITL (ms):                            185.14    
Max ITL (ms):                            333.13   
```

TRTLLM Sparse NSA
```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 16        
Successful requests:                     80        
Benchmark duration (s):                  58.32     
Total input tokens:                      29016     
Total input text tokens:                 29016     
Total generated tokens:                  17012     
Total generated tokens (retokenized):    16922     
Request throughput (req/s):              1.37      
Input token throughput (tok/s):          497.55    
Output token throughput (tok/s):         291.71    
Peak output token throughput (tok/s):    592.00    
Peak concurrent requests:                21        
Total token throughput (tok/s):          789.27    
Concurrency:                             11.27     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   8217.63   
Median E2E Latency (ms):                 6536.17   
P90 E2E Latency (ms):                    18150.71  
P99 E2E Latency (ms):                    33348.26  
---------------Time to First Token----------------
Mean TTFT (ms):                          659.34    
Median TTFT (ms):                        287.29    
P99 TTFT (ms):                           2945.19   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          37.39     
Median TPOT (ms):                        37.52     
P99 TPOT (ms):                           90.29     
---------------Inter-Token Latency----------------
Mean ITL (ms):                           35.82     
Median ITL (ms):                         27.18     
P95 ITL (ms):                            152.73    
P99 ITL (ms):                            189.19    
Max ITL (ms):                            2460.31 
```

Unit benchmarking shows higher bandwidth for the new trtllm kernel compared to flashmla.

bf16kvcache_comparison_bandwidth_vs_seqlen_bar

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) (`/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`) or contact authorized users to do so.
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/layers/attention/nsa_backend.py [hunk 1]
```diff
@@ -315,11 +317,12 @@ def __init__(
...
 
-        # Allocate global workspace buffer for TRTLLm ragged attention kernel (SM100/B200)
-        if self.device_sm_major >= 10:
+        # Allocate global workspace buffer for TRT-LLM kernels (ragged attention on SM100/B200, or trtllm_gen decode)
+        # Use torch.zeros() for zero-initialization required by TRT-LLM Gen multi-block reduction
+        if self.device_sm_major >= 10 or self.nsa_decode_impl == "trtllm_gen":
             global global_workspace_buffer
             if global_workspace_buffer is None:
-                global_workspace_buffer = torch.empty(
+                global_workspace_buffer = torch.zeros(
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/nsa_backend.py]: Why setting to `zeros`? It's slower than torch.empty

File: python/sglang/srt/server_args.py [hunk 2]
```diff
@@ -1111,6 +1116,20 @@ def _handle_model_specific_adjustments(self):
...
 
+                    user_set_prefill = self.nsa_prefill_backend is not None
+                    user_set_decode = self.nsa_decode_backend is not None
+
+                    # If user specified a backend but didn't explicitly set kv_cache_dtype,
+                    # require them to be explicit about kv_cache_dtype to avoid surprises
+                    if (
+                        user_set_prefill or user_set_decode
+                    ) and self.kv_cache_dtype == "auto":
+                        raise ValueError(
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py]: Please change ValueError to warning, since ValueError might break some prior usages

File: python/sglang/srt/server_args.py [hunk 3]
```diff
@@ -1126,17 +1145,26 @@ def _handle_model_specific_adjustments(self):
...
-                        # set prefill/decode backends to flashmla_sparse for Blackwell.
-                        # The default settings (P=flashmla_sparse, D=fa3) are for Hopper.
+                        # set prefill/decode backends to flashmla_sparse based on architecture.
                         if major >= 10:
-                            self.nsa_prefill_backend = "flashmla_sparse"
-                            self.nsa_decode_backend = "flashmla_sparse"
+                            if not user_set_prefill:
+                                self.nsa_prefill_backend = "flashmla_sparse"
+                            if not user_set_decode:
+                                self.nsa_decode_backend = "flashmla_sparse"
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py]: Can we set decode backend to trtllm_gen here by default, since it's faster?

File: python/sglang/srt/layers/attention/nsa_backend.py [hunk 4]
```diff
@@ -1701,6 +1715,41 @@ def _forward_aiter(
...
+
+        q = q_all.view(batch_size, 1, num_heads, head_dim)
+        kv = kv_cache.view(-1, 1, self.real_page_size, self.kv_cache_dim)
+        block_tables = page_table_1.unsqueeze(1)
+
+        out = flashinfer.decode.trtllm_batch_decode_with_kv_cache_mla(
+            query=q,
+            kv_cache=kv,
+            workspace_buffer=self.workspace_buffer,
+            qk_nope_head_dim=128,
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/nsa_backend.py]: Avoid hard coded numbers here. We had better replace them with model configs, or do some assertions on the dims.

**Review by Fridge003 — DISMISSED** [review]

**Comment by Fridge003** [comment]
@akhilg-nv  Please fix the conflict

File: python/sglang/srt/server_args.py (hunk 3 above)
Comment (akhilg-nv) [reply python/sglang/srt/server_args.py]: Would you like it to be the default if 1) the user doesn't specify kv cache dtype, and/or 2) bf16 kv cache is specified? The e2e performance can be better compared to flashmla with fp8 kvcache, but there is bandwidth tradeoffs. Right now I have set it to be the default only if bf16 kv cache is specified.

File: python/sglang/srt/layers/attention/nsa_backend.py (hunk 1 above)
Comment (akhilg-nv) [reply python/sglang/srt/layers/attention/nsa_backend.py]: https://github.com/flashinfer-ai/flashinfer/blob/885fbf53d70256db2040379ccb4c881baa624fe3/tests/attention/test_trtllm_gen_attention.py#L562-L564

This check indicates the workspace uses a counter and needs the initialized values to be zero. However, I tried comparing with empty initialization and the results seemed to be the same, so it's possible that these addresses are being zero'd by the kernel anyway. I think empty initialization could be relatively safe, let me know what you think.

File: python/sglang/srt/server_args.py (hunk 2 above)
Comment (akhilg-nv) [reply python/sglang/srt/server_args.py]: Done

File: python/sglang/srt/layers/attention/nsa_backend.py (hunk 4 above)
Comment (akhilg-nv) [reply python/sglang/srt/layers/attention/nsa_backend.py]: There are assertions in the kernel for these params, but I have updated it to draw the values from the model config.

File: python/sglang/srt/layers/attention/nsa_backend.py (hunk 1 above)
Comment (Fridge003) [reply python/sglang/srt/layers/attention/nsa_backend.py]: We can keep the empty here

File: python/sglang/srt/server_args.py (hunk 3 above)
Comment (Fridge003) [reply python/sglang/srt/server_args.py]: Yeah this is ok. Just set it by default when bf16 kv cache is specified

File: python/sglang/srt/layers/attention/nsa_backend.py (hunk 1 above)
Comment (akhilg-nv) [reply python/sglang/srt/layers/attention/nsa_backend.py]: Done

File: python/sglang/srt/layers/attention/nsa_backend.py [hunk 5]
```diff
@@ -254,7 +254,9 @@ def topk_transform(
             assert False, f"Unsupported {self.topk_transform_method = }"
 
 
-_NSA_IMPL_T: TypeAlias = Literal["flashmla_sparse", "flashmla_kv", "fa3", "tilelang"]
+_NSA_IMPL_T: TypeAlias = Literal[
+    "flashmla_sparse", "flashmla_kv", "fa3", "tilelang", "trtllm_gen"
```
Comment (hlu1) [inline python/sglang/srt/layers/attention/nsa_backend.py]:
```suggestion
    "flashmla_sparse", "flashmla_kv", "fa3", "tilelang", "trtllm"
```
Let's keep this consistent with "trtllm_mla" and "trtllm_mha". Please fix the rest of the usage of trtllm gen

File: python/sglang/srt/layers/attention/nsa_backend.py [hunk 6]
```diff
@@ -1713,6 +1729,41 @@ def _forward_aiter(
         # kv_cache = kv_cache.view(-1, 1, layer.head_dim)
         return o
 
+    def _forward_trtllm_gen(
```
Comment (hlu1) [inline python/sglang/srt/layers/attention/nsa_backend.py]:
```suggestion
    def _forward_trtllm(
```

File: python/sglang/srt/layers/attention/nsa_backend.py [hunk 7]
```diff
@@ -1713,6 +1729,41 @@ def _forward_aiter(
...
+        self,
+        q_all: torch.Tensor,
+        kv_cache: torch.Tensor,
+        page_table_1: torch.Tensor,
+        metadata: NSAMetadata,
+        sm_scale: float,
+    ) -> torch.Tensor:
+        """Forward using TRT-LLM Gen sparse MLA kernel."""
```
Comment (hlu1) [inline python/sglang/srt/layers/attention/nsa_backend.py]:
```suggestion
        """Forward using TRT-LLM sparse MLA kernel."""
```

File: python/sglang/srt/server_args.py [hunk 8]
```diff
@@ -1176,17 +1195,26 @@ def _handle_model_specific_adjustments(self):
...
+                            if not user_set_prefill:
+                                self.nsa_prefill_backend = "flashmla_sparse"
+                            if not user_set_decode:
+                                self.nsa_decode_backend = "trtllm_gen"
+                        else:
+                            # Hopper defaults for bfloat16
+                            if not user_set_prefill:
+                                self.nsa_prefill_backend = "flashmla_sparse"
+                            if not user_set_decode:
+                                self.nsa_decode_backend = "fa3"
```
Comment (hlu1) [inline python/sglang/srt/server_args.py]: It's better to wrap these into a function and then assign the return values to `self.nsa_prefill_backend` and `self.nsa_decode_backend`

```
def _get_default_nsa_backends(kv_cache_dtype):
    xxxx
    return nsa_prefill_backend, nsa_decode_backend

nsa_prefill_backend, nsa_decode_backend = _get_default_nsa_backends(kv_cache_dtype)

self.nsa_prefill_backend = self.nsa_prefill_backend or nsa_prefill_backend
self.nsa_decode_backend = self.nsa_decode_backend or nsa_decode_backend
```

Comment (akhilg-nv) [reply python/sglang/srt/server_args.py]: I've refactored the default kv cache dtype and nsa backend selection into separate setter methods. I chose to use setters instead of getters for two reasons:
1) it removes the need to account for cases where the user has specified exactly one nsa backend, in which case we wouldn't want to return both defaults
2) it seems more consistent with the pattern of other helpers in this file

File: python/sglang/srt/layers/attention/nsa_backend.py (hunk 5 above)
Comment (akhilg-nv) [reply python/sglang/srt/layers/attention/nsa_backend.py]: Done

**Review by hlu1 — APPROVED** [review]

File: python/sglang/srt/server_args.py [hunk 9]
```diff
@@ -1086,6 +1091,58 @@ def _generate_piecewise_cuda_graph_tokens(self):
 
         return capture_sizes
 
+    def _set_default_kv_cache_dtype(self, major: int) -> str:
```
Comment (hlu1) [inline python/sglang/srt/server_args.py]:
```suggestion
    def _set_default_nsa_kv_cache_dtype(self, major: int) -> str:
```
